src/components/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,464 @@
+import { Listbox } from '@headlessui/react';
+import clsx from 'clsx';
+import React, {
+  ReactNode,
+  ComponentProps,
+  ElementType,
+  useContext,
+} from 'react';
+import styles from './Dropdown.module.css';
+import DropdownButton from '../DropdownButton';
+import Icon from '../Icon';
+
+export type OptionsAlignType = 'left' | 'right';
+export type VariantType = 'compact' | 'full';
+
+type ListboxProps = ComponentProps<typeof Listbox>;
+type DropdownProps = ListboxProps & {
+  /**
+   * Text for the dropdown label.
+   *
+   * This is an alternative to passing <Dropdown.Label> in via `children`.
+   * If you pass in `labelText`, we expect `buttonText`, and `options` props
+   * as well and no `children`.
+   */
+  labelText?: ReactNode | string;
+  /**
+   * Screen-reader text for the dropdown label.
+   *
+   * When possible, use a visible label through the `labelText` prop or
+   * by passing a <Label> into `chidren`. In rare cases where there's no
+   * visible label, you must provide an `aria-label` for screen readers.
+   * If you pass in an `aria-label`, you don't need `labelText` or <Label>.
+   */
+  'aria-label'?: string;
+  /**
+   * Text for the dropdown label.
+   *
+   * This is an alternative to passing <Dropdown.Button> in via children.
+   * If you pass in `buttonText`, we expect `labelText` (or `aria-label`),
+   * and `options` props as well and no `children`.
+   */
+  buttonText?: ReactNode | string;
+  /**
+   * All options to be displayed in the dropdown list, passed in as
+   * an array of objects.
+   *
+   * This is an alternative to passing in the options via children.
+   * If you pass in `options`, we expect `labelText` (or `aria-label`),
+   * and `buttonText` props as well and no `children`.
+   */
+  options?: Array<{
+    key: string | number;
+    label: string;
+  }>;
+  /**
+   * Optional className for additional styling.
+   */
+  className?: string;
+  /**
+   * The style of the dropdown.
+   *
+   * Compact renders dropdown button that is only as wide as the content.
+   */
+  variant?: VariantType;
+  /**
+   * Align dropdown menu to the left (default) or right of dropdown button.
+   */
+  optionsAlign?: OptionsAlignType;
+  /**
+   * Optional className for additional options menu styling.
+   *
+   * When not using the compact variant, if optionsClassName is provided please
+   * include the width property to define the options menu width.
+   */
+  optionsClassName?: string;
+};
+
+type RenderProp<Arg> = (arg: Arg) => ReactNode;
+type PropsWithRenderProp<RenderPropArg> = {
+  children?: ReactNode | RenderProp<RenderPropArg>;
```

Comment:
This is for `<DropdownTrigger>`. Kept the `children` prop since more complex button triggers can be used, such as in ‘UsingFunctionChildrenProp’ story.